src/core/thread/indirect_sender.cpp

```diff
@@ -418,37 +418,51 @@ uint16_t IndirectSender::PrepareDataFrame(Mac::TxFrame &aFrame, Child &aChild, M
 
 void IndirectSender::PrepareEmptyFrame(Mac::TxFrame &aFrame, Child &aChild, bool aAckRequest)
 {
-    uint16_t     fcf;
-    Mac::Address macSource, macDest;
-
+    Mac::Address macDest;
     aChild.GetMacAddress(macDest);
+    PrepareEmptyFrame(aFrame, macDest, aAckRequest);
+}
 
+void IndirectSender::PrepareEmptyFrame(Mac::TxFrame &aFrame, const Mac::Address &aMacDest, bool aAckRequest)
+{
+    uint16_t fcf = 0;
+
+    Mac::Address macSource;
     macSource.SetShort(Get<Mac::Mac>().GetShortAddress());
 
-    if (macSource.IsShortAddrInvalid() || macDest.IsExtended())
+    if (macSource.IsShortAddrInvalid() || aMacDest.IsExtended())
     {
         macSource.SetExtended(Get<Mac::Mac>().GetExtAddress());
     }
 
-    fcf = Mac::Frame::kFcfFrameData | Mac::Frame::kFcfFrameVersion2006 | Mac::Frame::kFcfPanidCompression |
-          Mac::Frame::kFcfSecurityEnabled;
+    fcf = Mac::Frame::kFcfFrameData | Mac::Frame::kFcfPanidCompression | Mac::Frame::kFcfSecurityEnabled;
 
     if (aAckRequest)
     {
         fcf |= Mac::Frame::kFcfAckRequest;
     }
 
-    fcf |= (macDest.IsShort()) ? Mac::Frame::kFcfDstAddrShort : Mac::Frame::kFcfDstAddrExt;
+    fcf |= (aMacDest.IsShort()) ? Mac::Frame::kFcfDstAddrShort : Mac::Frame::kFcfDstAddrExt;
     fcf |= (macSource.IsShort()) ? Mac::Frame::kFcfSrcAddrShort : Mac::Frame::kFcfSrcAddrExt;
+    Get<Mac::Mac>().UpdateFrameControlField(nullptr, 0, fcf);
 
     aFrame.InitMacHeader(fcf, Mac::Frame::kKeyIdMode1 | Mac::Frame::kSecEncMic32);
 
-    aFrame.SetDstPanId(Get<Mac::Mac>().GetPanId());
-    IgnoreError(aFrame.SetSrcPanId(Get<Mac::Mac>().GetPanId()));
-    aFrame.SetDstAddr(macDest);
+    if (aFrame.IsDstPanIdPresent())
+    {
+        aFrame.SetDstPanId(Get<Mac::Mac>().GetPanId());
+    }
+    if (aFrame.IsSrcPanIdPresent())
+    {
+        IgnoreError(aFrame.SetSrcPanId(Get<Mac::Mac>().GetPanId()));
+    }
```

Comment:
Yes. In `SetSrcPanId` there's an assertion.

Reply:
I can see there is assert in `SetDstPanId`, but I fails to see it in `SetSrcPanId` in the master upstream.

Reply:
Before this PR, the empty data frame always uses short address so the assertion won't be triggered. I found the problem because I used Ext address for empty data frame in early time in this PR. Though I changed to use short address now, it's necessary to add the check (if we use this method again with extended address in the future).